### PR TITLE
INGEST: Create Index Before Pipeline Execute

### LINKED ISSUE #32758: [ingest] Default ingest node pipeline via index template fails to apply for first document.
A ingest pipeline defined as a default from an index template will not apply the pipeline to the very first document indexed.   

```
curl -XDELETE "http://localhost:9200/test"
curl -XPUT "http://localhost:9200/_ingest/pipeline/test-pipeline" -H 'Content-Type: application/json' -d'
{
  "description" : "test pipeline",
  "processors" : [
    {
      "set" : {
        "field" : "foo",
        "value" : "bar"
      }
    }
  ]
}'
curl -XPUT "http://localhost:9200/_template/template_test" -H 'Content-Type: application/json' -d'
{
    "index_patterns": ["test"],
    "settings": {
        "index.default_pipeline": "test-pipeline"
    }
}'
curl -XPOST "http://localhost:9200/test/_doc/1" -H 'Content-Type: application/json' -d'
{
  "a" : "b"
}'
curl -XPOST "http://localhost:9200/test/_refresh"
curl -XGET "http://localhost:9200/test/_search?pretty"
curl -XPOST "http://localhost:9200/test/_doc/2" -H 'Content-Type: application/json' -d'
{
  "a" : "b"
}'
curl -XPOST "http://localhost:9200/test/_refresh"
curl -XGET "http://localhost:9200/test/_search?pretty"
```

results in 
```
"hits" : {
    "total" : 2,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "a" : "b"
        }
      },
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 1.0,
        "_source" : {
          "a" : "b",
          "foo" : "bar"
        }
      }
    ]
  }
```

^^ note the the set pipeline did not take effect on the first document. 

It works fine when creating the index without without a template
```
curl -XDELETE "http://localhost:9200/test"
curl -XPUT "http://localhost:9200/_ingest/pipeline/test-pipeline" -H 'Content-Type: application/json' -d'
{
  "description" : "test pipeline",
  "processors" : [
    {
      "set" : {
        "field" : "foo",
        "value" : "bar"
      }
    }
  ]
}'
curl -XPUT "http://localhost:9200/test" -H 'Content-Type: application/json' -d'
{
    "settings" : {
        "index" : {
            "default_pipeline" : "test-pipeline" 
        }
    }
}'
curl -XPOST "http://localhost:9200/test/_doc/1" -H 'Content-Type: application/json' -d'
{
  "a" : "b"
}'
curl -XPOST "http://localhost:9200/test/_refresh"
curl -XGET "http://localhost:9200/test/_search?pretty"
curl -XPOST "http://localhost:9200/test/_doc/2" -H 'Content-Type: application/json' -d'
{
  "a" : "b"
}'
curl -XPOST "http://localhost:9200/test/_refresh"
curl -XGET "http://localhost:9200/test/_search?pretty"
```
output is as expected:
```
"hits" : {
    "total" : 2,
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "1",
        "_score" : 1.0,
        "_source" : {
          "a" : "b",
          "foo" : "bar"
        }
      },
      {
        "_index" : "test",
        "_type" : "_doc",
        "_id" : "2",
        "_score" : 1.0,
        "_source" : {
          "a" : "b",
          "foo" : "bar"
        }
      }
    ]
  }
```

cc: @original-brownbear

### PR DESCRIPTION
**Pull request by original-brownbear**
* Ensures that indices are created before the default pipeline setting is read to correcly handle the case of an index template containing a default pipeline (without the fix the first document does not get the pipeline applied as explained in #32758)
* Wasn't quite sure what a reasonable UT would be here, I can try writing one if we want it but it seems that would be a lot of mocking from where we are at with the test infra here (we are currently just mocking the cluster service, so we'd have to mock the steps for index creation + verify the right order of calls I guess? ... not sure if that's a valuable test but happy to add it :)).
* closes #32758

### CODE REVIEW

**Review by rjernst — CHANGES_REQUESTED** [review]
A unit test is preferable, and can fit in TransportBulkActionTests? I don't think there should be an integration test, as the behavior is completely internal to TransportBulkAction.

**Comment by original-brownbear** [comment]
@rjernst 

the problem with a UT (and as a result with leaving the IT out) is that we would basically want to test the the sequence of calling the transport client for setting up the new index, and then only loading the index+pipeline settings (and finally running the pipeline) once that index is up.

That's a lot of new test code that our current `org.elasticsearch.action.bulk.TransportBulkActionIngestTests` doesn't provide (it just hard uses a hard coded subclass of `TransportBulkAction` that has the auto create check overriden to false and uses `null` for the transport client).

... but I guess no time like the present to improve our test infra here :) On it.

**Comment by original-brownbear** [comment]
@rjernst how about this replacement for the IT:

https://github.com/elastic/elasticsearch/pull/32786/commits/37d28ca0b25bbdeab653457b20fb7069804c7e00

Implemented the index creation check in the same style used by `org.elasticsearch.action.bulk.TransportBulkActionTests.TestTransportBulkAction#createIndex`. Def. a little dirty to override the class we test for testing, but how about it? :)

File: server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java [hunk 1]
```diff
@@ -112,25 +120,42 @@
     /** A subclass of the real bulk action to allow skipping real bulk indexing, and marking when it would have happened. */
     class TestTransportBulkAction extends TransportBulkAction {
         boolean isExecuted = false; // set when the "real" bulk execution happens
+
+        boolean needToCheck;
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/action/bulk/TransportBulkActionIngestTests.java]: please add a comment that this makes the return of `needToCheck()` pluggable

**Review by rjernst — APPROVED** [review]
Thanks @original-brownbear. This will work I think for now.

**Comment by original-brownbear** [comment]
@rjernst thanks, comment added, will merge once green :)

**Pull request merged**